pyccel/codegen/compiling/file_locks.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/devel/LICENSE for full license details.      #
+#------------------------------------------------------------------------------------------#
+"""
+Module handling classes which handle file locking to avoid deadlocks.
+"""
+from filelock import FileLock
+
+class FileLockSet:
+    """
+    Class for grouping file locks.
+
+    A class which groups file locks. By grouping these the locking can
+    be handled via a context manager which reduces the risk of the locks
+    not being correctly released.
+
+    Parameters
+    ----------
+    locks : iterable[FileLock], optional
+        The locks that should be stored in the FileLockSet.
+    """
+    def __init__(self, locks = ()):
+        assert all(isinstance(l, FileLock) for l in locks)
+        self._locks = list(locks)
+
+    def __enter__(self):
+        for l in self._locks:
+            l.acquire()
+
+    def __exit__(self, *args):
```

Comment:
Shouldn't the `__exit__` method release the locks instead of acquiring them? A context manager is also supposed to handle exceptions to ensure that the locks will be released.

Reply:
:facepalm: I'm an idiot. I'm not sure how this didn't deadlock

Reply:
> A context manager is also supposed to handle exceptions to ensure that the locks will be released.

What do you mean by this? It is my understanding that this code will always be run even if an exception is raised (this is why context managers are preferred over the current strategy).
https://docs.python.org/3/reference/datamodel.html#object.__exit__

The context manager **may** decide to intercept the exception if it wants to handle this. In my case I don't want to intercept the exception, I am happy for it to be raised I just need to release the locks first

Reply:
I was thinking  about changing it to `def __exit__(self, exc_type, exc_val, exc_tb):` to reflect that it intercepts errors. However, you are right there is no need for it.

Reply:
Ah ok. I think it is better to be explicit with arguments. I didn't check the official docs when I wrote this which is how I ended up with `*args`. Fixed 7cde31fb